test/bufr/TestBufrParser.h

```diff
@@ -0,0 +1,158 @@
+
+#include <iomanip>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "eckit/config/LocalConfiguration.h"
+#include "eckit/testing/Test.h"
+#include "oops/runs/Test.h"
+#include "oops/util/Expect.h"
+#include "test/TestEnvironment.h"
+
+#include "BufrParser/BufrDescription.h"
+#include "BufrParser/BufrMnemonicSet.h"
+#include "BufrParser/BufrParser.h"
+#include "BufrParser/BufrTypes.h"
+#include "DataContainer.h"
+#include "ParserFactory.h"
+
+#include "DataObject/ArrayDataObject.h"
+#include "DataObject/StrVecDataObject.h"
+
+
+namespace Ingester
+{
+    namespace test
+    {
+        class BufrParserTestFixture : private boost::noncopyable
+        {
+         public:
+            static std::shared_ptr<Parser>& bufrParser() { return getInstance().bufrParser_; }
+
+         private:
+            std::shared_ptr<Parser> bufrParser_;
+
+            static BufrParserTestFixture& getInstance()
+            {
+                static BufrParserTestFixture bufrParserTestFixture;
+                return bufrParserTestFixture;
+            }
+
+            BufrParserTestFixture()
+            {
+                registerParser();
+
+                const eckit::LocalConfiguration conf(::test::TestEnvironment::config());
+
+                if (conf.has("observations"))
+                {
+                    auto obsConf = conf.getSubConfigurations("observations").front();
+
+                    if (obsConf.has("obs space") &&
+                        obsConf.getSubConfiguration("obs space").has("name") &&
+                        obsConf.getSubConfiguration("obs space").getString("name") == "bufr")
+                    {
+                        auto bufrConf = obsConf.getSubConfiguration("obs space");
+                        bufrParser_ = ParserFactory::create(bufrConf);
+                    }
+                    else
+                    {
+                        throw eckit::BadValue(
+                            "Configuration File is missing the \"bufr\" section.");
+                    }
+                }
+                else
+                {
+                    throw eckit::BadValue(
+                        "Configuration File is missing the \"observations\" section.");
+                }
+            }
+
+            void registerParser()
+            {
+                ParserFactory::registerParser<BufrParser>("bufr");
+            }
+
+            ~BufrParserTestFixture() = default;
+        };
+
+
+        void test_constructor()
+        {
+            BufrParserTestFixture::bufrParser();
+        }
+
+        void test_parsePartialFile()
+        {
+            auto data = BufrParserTestFixture::bufrParser()->parse(5);
+            auto dataObj = data->get("radiance");
+
+            if (auto arrayDataObject = std::dynamic_pointer_cast<ArrayDataObject>(dataObj))
+            {
+                EXPECT(abs(arrayDataObject->get()(0, 0) - 248.17) < .01);
```

Comment:
Can you make this expect thing be a YAML configurable option? So that we can update the test input file and not have to change the source code.

Also, can we make a better test than just if the first value is correct? What about a mean of all the data within some tolerance? Or min and max? Just thinking out loud here, but I'd like to see a better test. Alternatively, you can remove this entirely and determine if the test passes by if the output matches a reference output file using nccmp.

Reply:
All good points.

Reply:
Coming up with a good scheme for this will require a little thought. We are not using this unit test at the moment so I think I'm going to defer action on this comment. I think we will probably do a unit test branch in the future and merge those changes as a seperate PR..